# Anwendung der .groupby() Funktion in Python

Dieses Notebook dient der Anwendung der groupby Funktion innerhalb verschiedener Datensätze. Es wird stetig erweitert, um das Wissen und Neuerungen aktuell zu halten. Die Aufgaben sind in leicht, mittel und schwer aufgeteilt und beziehen sich auf das Dokument [car_insurance_claims.csv](https://github.com/EikeSuemper/Meine-Lernreise-Aktuariat-QuantFinance/blob/main/04_Datens%C3%A4tze/car_insurance_claims.csv)

In [1]:
import numpy as np
import pandas as pd

df1 = pd.read_csv("https://raw.githubusercontent.com/EikeSuemper/Meine-Lernreise-Aktuariat-QuantFinance/refs/heads/main/04_Datens%C3%A4tze/car_insurance_claims.csv")
df1.head()

,months_as_customer,age,policy_number,policy_bind_date,policy_state,policy_csl,policy_deductable,policy_annual_premium,umbrella_limit,insured_zip,...,police_report_available,total_claim_amount,injury_claim,property_claim,vehicle_claim,auto_make,auto_model,auto_year,fraud_reported,_c39
0,328,48,521585,2014-10-17,OH,250/500,1000,1406.91,0,466132,...,YES,71610,6510,13020,52080,Saab,92x,2004,Y,NaN
1,228,42,342868,2006-06-27,IN,250/500,2000,1197.22,5000000,468176,...,?,5070,780,780,3510,Mercedes,E400,2007,Y,NaN
2,134,29,687698,2000-09-06,OH,100/300,2000,1413.14,5000000,430632,...,NO,34650,7700,3850,23100,Dodge,RAM,2007,N,NaN
3,256,41,227811,1990-05-25,IL,250/500,2000,1415.74,6000000,608117,...,NO,63400,6340,6340,50720,Chevrolet,Tahoe,2014,Y,NaN
4,228,44,367455,2014-06-06,IL,500/1000,1000,1583.91,6000000,610706,...,NO,6500,1300,650,4550,Accura,RSX,2009,N,NaN


## Leichte Aufgaben

### Aufgabe 1

Gruppiere den DataFrame nach policy_state und berechne die Anzahl der Schadensfälle pro Bundesstaat.

In [2]:
df1.groupby("policy_state").size()

policy_state
IL    338
IN    310
OH    352
dtype: int64

In [3]:
ergebnis1 = df1.groupby("policy_state").size().reset_index(name="Anzahl_Faelle")
ergebnis1.sort_values('Anzahl_Faelle', ascending=False)

,policy_state,Anzahl_Faelle
2,OH,352
0,IL,338
1,IN,310


Mit .size() werden die Anzahl der Zeilen je Staat gezählt und mit reset_index() die Serie in ein DataFrame gewandelt und mit der Spalte "Anzahl Fälle" aufgefüllt.

Es gibt 352 Schadenfälle in Ohio, 338 Schadenfälle in Illinois und 310 Schadenfälle in Indiana.

### Aufgabe 2

Gruppiere nach incident_severity und zeige die durchschnittliche Schadenssumme (total_claim_amount) für jeden Schweregrad.

In [4]:
ergebnis2 = df1.groupby("incident_severity")["total_claim_amount"].mean().round(2)
ergebnis2.reset_index(name="Durchschnitt_Schaden_EUR")

,incident_severity,Durchschnitt_Schaden_EUR
0,Major Damage,64067.17
1,Minor Damage,48642.68
2,Total Loss,62081.21
3,Trivial Damage,5301.67


### Aufgabe 3

Gruppiere nach insured_sex und berechne das durchschnittliche Alter (age) der Versicherten.

In [5]:
ergebnis3 = df1.groupby('insured_sex')['age'].mean().round(2).reset_index(name='Durchschnitt_Alter')
ergebnis3

,insured_sex,Durchschnitt_Alter
0,FEMALE,38.33
1,MALE,39.67


### Aufgabe 4

Gruppiere nach auto_make (Automarke) und zähle, wie viele Schadensfälle es pro Marke gibt.

In [6]:
ergebnis4 = df1.groupby("auto_make").size().reset_index(name="anzahl_schadensfaelle")
ergebnis4

,auto_make,anzahl_schadensfaelle
0,Accura,68
1,Audi,69
2,BMW,72
3,Chevrolet,76
4,Dodge,80
5,Ford,72
6,Honda,55
7,Jeep,67
8,Mercedes,65
9,Nissan,78


### Aufgabe 5

Gruppiere nach insured_sex und berechne den Mittelwert von umbrella_limit.

In [7]:
ergebnis5 = df1.groupby("insured_sex")["umbrella_limit"].mean().reset_index(name="durchschnitt_umbrella_limit")
ergebnis5

,insured_sex,durchschnitt_umbrella_limit
0,FEMALE,1.078212e+06
1,MALE,1.127430e+06
